[quantization] Move quantized::mul_scalar to ATen

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27258 [quantization] Move quantized::mul_scalar to ATen**
* #27257 [quantization] Move quantized::add_scalar to ATen

Differential Revision: [D17731565](https://our.internmc.facebook.com/intern/diff/D17731565)

Commit:
[quantization] Move quantized::mul_scalar to ATen

Commit:
Update on "[quantization] Move quantized::mul_scalar to ATen"

Commit:
Update on "[quantization] Move quantized::mul_scalar to ATen"

Commit:
Update on "[quantization] Move quantized::mul_scalar to ATen"


Differential Revision: [D17731565](https://our.internmc.facebook.com/intern/diff/D17731565)

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -173,27 +173,48 @@ static auto registry =
+// ATen bindings for mul
+
+Tensor & quantized_mul_out(Tensor & out, const Tensor & self, const Tensor & other) {
+  TORCH_CHECK(other.sizes().size() == 0, kTensorTensorError);
```
this error message is about torchadd which seems wrong

dzhulgakov commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1956,17 +1957,33 @@
 
+- func: mul_relu.Tensor(Tensor self, Tensor other) -> Tensor
+  use_c10_dispatcher: full
+  variants: function, method
```
can you not make it a function? and maybe make the namespace for the function to be 'python_module: nn' (then it'd be torch._C._nn.mul_relu? I don't think we should expose `torch.mul_relu` as top level function and definitely not `Tensor.mul_relu`

raghuramank100 replied:
We do not need to rush this for 1.3, scalar add and mul do occur in a few models, but I havent yet seen a mul_relu or an add_relu for scalars yet.

To be cleaner, should we also have a float variant of this function so that the API does not diverge between float and quantized? 
Also agree that we should not expose fused operations in torch. namespace

dzhulgakov commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -148,7 +148,7 @@ def mul(self, x, y):
     r"""Operation equivalent to ``torch.ops.quantized.mul(Tensor, float)``"""
     def mul_scalar(self, x, y):
```
I don't even think we need mul_scalar now - the invocation will do dispatch, no?

dzhulgakov replied:
ah, nevermind, it's because of different args - passing or not qscale and qbias. Maybe just hide them in the same php method and do the check on numel inside the method? (but I'm not insisting, at least add it as a comment)

dzhulgakov commented on torch/nn/quantized/modules/functional_modules.py:
```diff
@@ -148,7 +148,7 @@ def mul(self, x, y):
+        return x * y
 
     r"""Operation equivalent to ``torch.ops.quantized.cat``"""
     def cat(self, x, dim=0):
```
mul_relu actually should be exposed here too

Commit:
Update on "[quantization] Move quantized::mul_scalar to ATen"


Differential Revision: [D17731565](https://our.internmc.facebook.com/intern/diff/D17731565)

Commit:
Update on "[quantization] Move quantized::mul_scalar to ATen"


Differential Revision: [D17731565](https://our.internmc.facebook.com/intern/diff/D17731565)

zafartahirov commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -173,27 +173,48 @@ static auto registry =
 }  // namespace
+
+constexpr const char *kTensorTensorError = "torch.mul is not supported when multiplying two"
+" quantized tensors. Please use torch.nn.quantized.modules.FloatFunctional";
```
I don't think we should be giving suggestions, in case `FloatFunctional` is replaced/renamed.

z-a-f reviewed (approved):
LGTM
